### PR TITLE
Issue 6155: Secure Access to Segment Store CLI

### LINKED ISSUE #6155: Secure mode for Segment Store CLI Commands
**Summary**
The segment store admin cli commands requires option to enable security such as authentication and TLS. This will allow it to be used in secured environments such as SDP.

**Describe the solution you'd like**
Add `delegation token` when dealing with wire commands related to each of segment store cli commands. `GrpcAuthHelper` can be used in `SegmentStoreCommand` to create a delegation token for `GetSegmentAttributCommand`, `GetSegmentInfoCommand`, `ReadSegmentRangeCommand`, and `UdpateSegmentAttributeCommand`.

The `SegmentStoreCommandTest` requires modification to test segment store commands in secured environment.

### PR DESCRIPTION
**Pull request by dellThejas**
Change log description

Added `delegationtoken` for segment store cli commands.
Updated unit tests for testing segment store cli commands with tls and auth.

Purpose of the change
Fixes #6155 

What the code does

- Added `GrpcAuthHelper` in `SegmentstoreCommand` to create a delegation token for all segment store commands
- `GetSegmentAtrributeCommand`, `GetSegmentInfoCommand` and `ReadSegmentRangeCommand`, `UpdateSegmentAttributeCommand` are now passing in delegationToken to allow the cli commands to run with security enabled.
- Updated `SegmentStoreCommandTest` to `AbstractSegmentStoreCommandsTest` with option to test segment store cli commands with security.

How to verify it

- Test using pravega standalone with tls and auth enabled.
- Run `AbstractSegmentStoreCommandsTest` for unit tests.

### CODE REVIEW

**Comment by RaulGracia** [comment]
@dellThejas you have multiple problems in this PR, like i) improper documentation/structure of the PR description, ii) missing an associated issue that describes the problem to solve, iii) failing checks that prevent the build to complete. Please, review this document https://github.com/pravega/pravega/wiki/Contributing and try to execute `./gradlew check` locally to fix build issues faster.

**Comment by dellThejas** [comment]
Hi @RaulGracia, I made this only to make it easier to share changes while I am debugging. I will make sure the documentation and everything is updated before I request a review.

File: cli/admin/README.md [hunk 1]
```diff
@@ -65,18 +65,19 @@ Pravega Admin CLI.
    
 Initial configuration:
     cli.store.metadata.backend=segmentstore
-    cli.controller.connect.credentials.username=admin
+    cli.credentials.username=admin
```
Comment (andreipaduroiu) [inline cli/admin/README.md:68]: Why are we changing these?

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/SegmentStoreCommand.java [hunk 2]
```diff
@@ -14,17 +14,22 @@
...
  */
 public abstract class SegmentStoreCommand extends AdminCommand {
     static final String COMPONENT = "segmentstore";
+    protected final GrpcAuthHelper authHelper;
 
     public SegmentStoreCommand(CommandArgs args) {
         super(args);
+
+        authHelper = new GrpcAuthHelper(true,
+                "secret",
```
Comment (andreipaduroiu) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/SegmentStoreCommand.java:32]: What is the purpose of these hardcoded values? Would we ever want to change them at runtime?

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIControllerConfig.java [hunk 3]
```diff
@@ -32,29 +32,30 @@
 
     private static final Property<String> CONTROLLER_REST_URI = Property.named("controller.connect.rest.uri", "localhost:9091");
     private static final Property<String> CONTROLLER_GRPC_URI = Property.named("controller.connect.grpc.uri", "localhost:9090");
-    private static final Property<Boolean> AUTH_ENABLED = Property.named("controller.connect.channel.auth", false);
-    private static final Property<String> CONTROLLER_USER_NAME = Property.named("controller.connect.credentials.username", "");
-    private static final Property<String> CONTROLLER_PASSWORD = Property.named("controller.connect.credentials.pwd", "");
-    private static final Property<Boolean> TLS_ENABLED = Property.named("controller.connect.channel.tls", false);
-    private static final Property<String> TRUSTSTORE_JKS = Property.named("controller.connect.trustStore.location", "");
+    private static final Property<Boolean> AUTH_ENABLED = Property.named("channel.auth", true);
```
Comment (andreipaduroiu) [inline cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIControllerConfig.java:35]: I see this is where the config name changes come from. Why did you remove `controller` out of these? For example `CONTROLLER_USER_NAME` still has `CONTROLLER` in it, but the property name has that `controller` removed from it.

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIControllerConfig.java [hunk 4]
```diff
@@ -97,6 +104,7 @@ private CLIControllerConfig(TypedProperties properties) throws ConfigurationExce
         this.userName = properties.get(CONTROLLER_USER_NAME);
         this.password = properties.get(CONTROLLER_PASSWORD);
         this.truststore = properties.get(TRUSTSTORE_JKS);
+        this.accessTokenTtlInSeconds = properties.getInt(TRUSTSTORE_ACCESS_TOKEN_TTL_SECONDS);
```
Comment (andreipaduroiu) [inline cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIControllerConfig.java:107]: The best practice (at least in our codebase), is to store all time durations as a `Duration` object. Check out other config objects in our codebase (like `ContainerConfig`, `WriterConfig`, etc) to see how to use those. 

I even believe that `TypedProperties` has a `getDuration` method to help you out.

File: test/integration/src/main/java/io/pravega/test/integration/demo/ClusterWrapper.java [hunk 5]
```diff
@@ -141,6 +141,10 @@
     @Getter
     private String tlsServerKeystorePasswordPath;
 
+    @Getter
+    @Builder.Default
+    private int accessTokenTtlInSeconds = 300;
```
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/demo/ClusterWrapper.java:146]: I suggest using `Duration`. It makes using it so much more cleaner.

File: cli/admin/README.md (hunk 1 above)
Comment (dellThejas) [reply cli/admin/README.md:68]: I made the change to make the use of credentials for pravega cli as a whole rather than only the controller in the pravega cli. Is that ok?

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/SegmentStoreCommand.java (hunk 2 above)
Comment (dellThejas) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/SegmentStoreCommand.java:32]: @anirudhkovuru mentioned that this value should not be left for the user to manipulate, because its a secure value that should be consistent. So, I took it off from the parameters we pass in for the cli. You can correct me if I missed anything Anirudh, ty.

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIControllerConfig.java (hunk 3 above)
Comment (dellThejas) [reply cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIControllerConfig.java:35]: I changed the it from `CONTROLLER_USER_NAME` to `USER_NAME`, and same for the password now.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/SegmentStoreCommand.java (hunk 2 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/SegmentStoreCommand.java:32]: In this case, `"secret"` refers to the `tokenSigningKey` which is a server secret that is used for controller and segmentstore communications.

Comment (andreipaduroiu) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/SegmentStoreCommand.java:32]: I see; so it's not actually the secret, rather the key that use used to get the actual secret from the environment, correct?

File: cli/admin/README.md (hunk 1 above)
Comment (andreipaduroiu) [reply cli/admin/README.md:68]: Ok. I eventually realized as I was reading down the PR last time.

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIConfig.java [hunk 6]
```diff
@@ -66,37 +70,44 @@
...
+     * Truststore if TLS is configured in the CLI.
      */
     @Getter
     private final String truststore;
 
+    /**
+     * Truststore access token ttl in seconds if TLS is configured in the CLI.
+     */
+    @Getter
+    private final Duration accessTokenTtlInSeconds;
```
Comment (andreipaduroiu) [inline cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIConfig.java:94]: rename this to "accessTokenTtl". `Duration` holds time, and you can use it to express it in any time unit you want.

Also fix the Javadoc please.

File: test/integration/src/main/java/io/pravega/test/integration/demo/ClusterWrapper.java [hunk 7]
```diff
@@ -141,6 +142,10 @@
     @Getter
     private String tlsServerKeystorePasswordPath;
 
+    @Getter
+    @Builder.Default
+    private Duration accessTokenTtlInSeconds = Duration.ofSeconds(300);
```
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/demo/ClusterWrapper.java:147]: remove "Seconds" from the field name.

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]
Looks good. Two very small comments, then we should be good to merge.

File: cli/admin/README.md [hunk 8]
```diff
@@ -77,7 +77,7 @@ Initial configuration:
     cli.controller.connect.rest.uri=localhost:9091
     cli.controller.connect.grpc.uri=localhost:9090
     cli.trustStore.location=conf/ca-cert.crt
-    cli.trustStore.access.token.ttl.seconds=300
+    cli.trustStore.access.token.ttl=300
```
Comment (andreipaduroiu) [inline cli/admin/README.md:80]: This string must contain the unit. It was OK the way it was before (with `.seconds` suffix).

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIConfig.java [hunk 9]
```diff
@@ -40,7 +40,7 @@
     private static final Property<String> PASSWORD = Property.named("credentials.pwd", "");
     private static final Property<Boolean> TLS_ENABLED = Property.named("channel.tls", false);
     private static final Property<String> TRUSTSTORE_JKS = Property.named("trustStore.location", "");
-    private static final Property<Integer> TRUSTSTORE_ACCESS_TOKEN_TTL_SECONDS = Property.named("trustStore.access.token.ttl.seconds", 10);
+    private static final Property<Integer> TRUSTSTORE_ACCESS_TOKEN_TTL = Property.named("trustStore.access.token.ttl", 10);
```
Comment (andreipaduroiu) [inline cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIConfig.java:43]: The **field** name is fine (no time unit). The property name (in the string) must have unit (`.seconds`).

File: cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIConfig.java [hunk 10]
```diff
@@ -91,7 +91,7 @@
      * Truststore access token ttl in seconds if TLS is configured in the CLI.
      */
     @Getter
-    private final Duration accessTokenTtlInSeconds;
+    private final Duration accessTokenTtl;
```
Comment (andreipaduroiu) [inline cli/admin/src/main/java/io/pravega/cli/admin/utils/CLIConfig.java:94]: This name is fine. This was the only change I asked you to do with my previous review's comment.

File: cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java [hunk 11]
```diff
@@ -137,11 +137,11 @@ public static ClusterWrapper createPravegaCluster(boolean authEnabled, boolean t
      * @param containerCount the container count.
      * @param authEnabled whether the cli requires authentication to access the cluster.
      * @param tlsEnabled whether the cli requires TLS to access the cluster.
-     * @param accessTokenTtlInSeconds how long the access token will last
+     * @param accessTokenTtl how long the access token will last
      */
     @SneakyThrows
     public static AdminCommandState createAdminCLIConfig(String controllerRestUri, String controllerUri, String zkConnectUri,
-                                                         int containerCount, boolean authEnabled, boolean tlsEnabled, Duration accessTokenTtlInSeconds) {
+                                                         int containerCount, boolean authEnabled, boolean tlsEnabled, Duration accessTokenTtl) {
```
Comment (andreipaduroiu) [inline cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java:144]: This is a good change.

File: config/admin-cli.properties [hunk 12]
```diff
@@ -20,7 +20,7 @@ cli.channel.tls=true
 cli.credentials.username=admin
 cli.credentials.pwd=1111_aaaa
 cli.trustStore.location=conf/ca-cert.crt
-cli.trustStore.access.token.ttl.seconds=600
+cli.trustStore.access.token.ttl=600
```
Comment (andreipaduroiu) [inline config/admin-cli.properties:23]: revert this.

File: test/integration/src/main/java/io/pravega/test/integration/demo/ClusterWrapper.java [hunk 13]
```diff
@@ -144,7 +144,7 @@
 
     @Getter
     @Builder.Default
-    private Duration accessTokenTtlInSeconds = Duration.ofSeconds(300);
+    private Duration accessTokenTtl = Duration.ofSeconds(300);
```
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/demo/ClusterWrapper.java:147]: This is a good change.

File: test/integration/src/main/java/io/pravega/test/integration/demo/ControllerWrapper.java [hunk 14]
```diff
@@ -97,11 +97,11 @@ public ControllerWrapper(final String connectionString, final boolean disableEve
                              final int controllerPort, final String serviceHost, final int servicePort,
                              final int containerCount, int restPort,
                              boolean enableAuth, String passwordAuthHandlerInputFilePath,
-                             String tokenSigningKey, int accessTokenTtlInSeconds) {
+                             String tokenSigningKey, int accessTokenTtl) {
```
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/demo/ControllerWrapper.java:100]: Put `seconds` back. You do not have a `Duration` that reflects time; you have an `int` where you need to tell the caller somehow how to interpret that value. Adding the `Seconds` suffix indicates that this should be time expressed in seconds.

File: test/integration/src/main/java/io/pravega/test/integration/demo/ControllerWrapper.java [hunk 15]
```diff
@@ -110,10 +110,10 @@ public ControllerWrapper(final String connectionString, final boolean disableEve
                              final int containerCount, int restPort,
                              boolean enableAuth, String passwordAuthHandlerInputFilePath,
                              String tokenSigningKey, boolean isRGWritesWithReadPermEnabled,
-                             int accessTokenTtlInSeconds) {
+                             int accessTokenTtl) {
```
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/demo/ControllerWrapper.java:113]: seconds

File: test/integration/src/main/java/io/pravega/test/integration/demo/ControllerWrapper.java [hunk 16]
```diff
@@ -123,7 +123,7 @@ public ControllerWrapper(final String connectionString, final boolean disableEve
                              final int containerCount, int restPort,
                              boolean enableAuth, String passwordAuthHandlerInputFilePath,
                              String tokenSigningKey, boolean isRGWritesWithReadPermEnabled,
-                             int accessTokenTtlInSeconds, boolean enableTls, String serverCertificatePath,
+                             int accessTokenTtl, boolean enableTls, String serverCertificatePath,
```
Comment (andreipaduroiu) [inline test/integration/src/main/java/io/pravega/test/integration/demo/ControllerWrapper.java:126]: seconds

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]

File: test/integration/src/main/java/io/pravega/test/integration/utils/SetupUtils.java [hunk 17]
```diff
@@ -86,32 +92,76 @@
...
     }
-    
+
     /**
      * Start all pravega related services required for the test deployment.
      *
      * @param numThreads the number of threads for the internal client threadpool.
      * @throws Exception on any errors.
      */
     public void startAllServices(Integer numThreads) throws Exception {
```
Comment (RaulGracia) [inline test/integration/src/main/java/io/pravega/test/integration/utils/SetupUtils.java:119]: Why is this needed?

File: cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java [hunk 18]
```diff
@@ -390,11 +390,21 @@ public SegmentHelper instantiateSegmentHelper(CuratorFramework zkClient) {
                 .containerCount(getServiceConfig().getContainerCount())
                 .build();
         HostControllerStore hostStore = HostStoreFactory.createStore(hostMonitorConfig, StoreClientFactory.createZKStoreClient(zkClient));
-        ClientConfig clientConfig = ClientConfig.builder()
-                .controllerURI(URI.create(getCLIControllerConfig().getControllerGrpcURI()))
-                .validateHostName(getCLIControllerConfig().isAuthEnabled())
-                .credentials(new DefaultCredentials(getCLIControllerConfig().getPassword(), getCLIControllerConfig().getUserName()))
-                .build();
+
+        ClientConfig.ClientConfigBuilder clientConfigBuilder = ClientConfig.builder()
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java:394]: The previous approach for building the configuration seems more concise. Please, try to keep it similar to what it was if possible, while setting the right values.

File: cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java [hunk 19]
```diff
@@ -390,11 +390,21 @@ public SegmentHelper instantiateSegmentHelper(CuratorFramework zkClient) {
...
+                .controllerURI(URI.create(getCLIControllerConfig().getControllerGrpcURI()));
+
+        if (getCLIControllerConfig().isAuthEnabled()) {
+            clientConfigBuilder.credentials(new DefaultCredentials(getCLIControllerConfig().getPassword(),
+                    getCLIControllerConfig().getUserName()));
+        }
+        if (getCLIControllerConfig().isTlsEnabled()) {
+            clientConfigBuilder.trustStore(getCLIControllerConfig().getTruststore())
+                    .validateHostName(false);
```
Comment (RaulGracia) [inline cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java:403]: Why is this hardcoded? Should this be yet another parameter?

**Review by RaulGracia — COMMENTED** [review]
Looks good, just some minor comments. One important point though: please, add a subsection in the Admin CLI readme with some instructions/guidelines on how to use the Admin CLI with security enabled.

File: cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java (hunk 18 above)
Comment (dellThejas) [reply cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java:394]: I dont think I can keep it the same, I am doing this approach based on other examples. I dont think we should be passing in some of those values into ClientConfig if tls or auth isnt enabled.

File: cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java (hunk 19 above)
Comment (dellThejas) [reply cli/admin/src/main/java/io/pravega/cli/admin/AdminCommand.java:403]: This value wont be changed. For example in admin/utils/TestUtils.java

File: cli/admin/README.md [hunk 20]
```diff
@@ -241,5 +242,17 @@ the appropriate implementation in `AdminRequestProcessorImpl`.
 - The last step would be to add each new `segmentstore` command as a separate class in the package where the existing
 commands are already placed.
 
+## Enable TLS and Auth in CLI
+Make sure to update the following fields in the configuration to enable tls ad auth in the cli:
```
Comment (RaulGracia) [inline cli/admin/README.md:246]: Nit: TLS and CLI in caps.

**Review by RaulGracia — APPROVED** [review]
LGTM, just one minor comment. I assume that you have tested this with security enabled in Pravega Standalone as well as a real Pravega cluster deployed with security enabled.

File: cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java [hunk 21]
```diff
@@ -131,22 +136,24 @@ public static ClusterWrapper createPravegaCluster(boolean authEnabled, boolean t
...
-        pravegaProperties.setProperty("cli.controller.connect.credentials.username", SecurityConfigDefaults.AUTH_ADMIN_USERNAME);
-        pravegaProperties.setProperty("cli.controller.connect.credentials.pwd", SecurityConfigDefaults.AUTH_ADMIN_PASSWORD);
-        pravegaProperties.setProperty("cli.controller.connect.channel.tls", Boolean.toString(tlsEnabled));
-        pravegaProperties.setProperty("cli.controller.connect.trustStore.location", pathToConfig() + SecurityConfigDefaults.TLS_CA_CERT_FILE_NAME);
+        pravegaProperties.setProperty("cli.channel.auth", Boolean.toString(authEnabled));
+        pravegaProperties.setProperty("cli.credentials.username", SecurityConfigDefaults.AUTH_ADMIN_USERNAME);
+        pravegaProperties.setProperty("cli.credentials.pwd", SecurityConfigDefaults.AUTH_ADMIN_PASSWORD);
+        pravegaProperties.setProperty("cli.channel.tls", Boolean.toString(tlsEnabled));
+        pravegaProperties.setProperty("cli.trustStore.location", pathToConfig() + SecurityConfigDefaults.TLS_CA_CERT_FILE_NAME);
+        pravegaProperties.setProperty("cli.trustStore.access.token.ttl", Long.toString(accessTokenTtl.toSeconds()));
```
Comment (andreipaduroiu) [inline cli/admin/src/test/java/io/pravega/cli/admin/utils/TestUtils.java:156]: You are not setting the correct property here. The property, as defined in `CLIConfig` is named `cli.trustStore.access.token.ttl.seconds`.

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]

**Review by andreipaduroiu — APPROVED** [review]
